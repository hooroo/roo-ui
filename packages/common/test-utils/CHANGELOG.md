# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

<a name="0.4.0"></a>
# [0.4.0](https://github.com/hooroo/roo-ui/compare/v0.3.3...v0.4.0) (2018-04-26)


### Features

* add theming ([bee896a](https://github.com/hooroo/roo-ui/commit/bee896a))
