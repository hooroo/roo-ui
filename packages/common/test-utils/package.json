{
  "name": "@roo-ui/test-utils",
  "version": "0.0.0",
  "main": "dist/index.js",
  "publishConfig": {
    "access": "public"
  },
  "scripts": {
    "build": "babel src -d dist --copy-files"
  }
}
