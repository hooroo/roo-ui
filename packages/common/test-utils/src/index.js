export { default as shallowWithTheme } from './shallowWithTheme'; // eslint-disable-line import/prefer-default-export
