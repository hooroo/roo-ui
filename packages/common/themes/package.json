{
  "name": "@roo-ui/themes",
  "version": "0.4.0",
  "main": "dist/index.js",
  "publishConfig": {
    "access": "public"
  },
  "scripts": {
    "build": "babel src -d dist --copy-files"
  }
}
