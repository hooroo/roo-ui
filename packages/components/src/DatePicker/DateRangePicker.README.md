# DateRangePicker

<!-- STORY -->

## Installation

```shell
$ yarn add @roo-ui/components
```

## Example

```js
import { DateRangePicker } from '@roo-ui/components';

export default (
  <DateRangePicker
    onRangeSelected={console.log}
  />
);
```

## Properties

| Name                       | Description                                      | Type      | Default | Required? |
|----------------------------|--------------------------------------------------|-----------|---------|-----------|
| `monthsToDisplay`          | Number of months to display                      | `number`  | 1       | -         |
| `stacked`                  | Stack the calendar months vertically             | `boolean` | false   | -         |
| `monthNames`               | Month names                                      | `array`   | -       | -         |
| `weekdayNames`             | Weekday names                                    | `array`   | -       | -         |
| `disabledDates`            | Dates which will be disabled                     | `array`   | []      | -         |
| `interactiveDisabledDates` | Disabled dates will be clickable                 | `boolean` | false   | -         |
| `startDate`                | Start date of the range                          | `date`    | -       | -         |
| `endDate`                  | End date of the range                            | `date`    | -       | -         |
| `setStartDate`             | Change the start date of the range               | `boolean` | false   | -         |
| `setEndDate`               | Change the end date of the range                 | `boolean` | false   | -         |
| `onRangeSelected`          | Callback function when the range is selected     | `boolean` | -       | true      |
| `onChangeStartDate`        | Callback function when the start date is chnaged | `boolean` | -       | true      |
| `onChangeEndDate`          | Callback function when the end date is chnaged   | `boolean` | -       | true      |

## Customization

This component can be passed additional props defined in the [dayzed](https://github.com/deseretdigital/dayzed) documentation.