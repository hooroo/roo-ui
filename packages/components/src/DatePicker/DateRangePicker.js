import React from 'react';
import PropTypes from 'prop-types';
import Dayzed from 'dayzed';
import { themeGet } from 'styled-system';
import { subDays, isEqual, isWithinRange } from 'date-fns';

import { Flex, Box, Input } from '../';

import CalendarNav from './components/CalendarNav';
import CalendarMonth from './components/CalendarMonth';

const DateInput = Input.extend`
  border: ${themeGet('borders.2')} ${themeGet('colors.greys.porcelain')};
  color: transparent;
  text-shadow: 0 0 0 ${themeGet('colors.greys.charcoal')};
  border-color: ${props => ((props.highlight) ? themeGet('colors.brand.secondary') : themeGet('colors.greys.porcelain'))};
`;

DateInput.defaultProps = {
  ...Input.defaultProps,
  blacklist: [...Object.keys(Input.propTypes), 'highlight'],
};

class DateRangePicker extends React.Component {
  constructor(props) {
    super(props);

    this.onMouseEnterOfDay = this.onMouseEnterOfDay.bind(this);
    this.onDateSelected = this.onDateSelected.bind(this);
    this.isInRange = this.isInRange.bind(this);

    this.state = {
      hoveredDate: null,
      startDate: this.props.startDate,
      endDate: this.props.endDate,
      isSettingStartDate: this.props.setStartDate,
      isSettingEndDate: this.props.setEndDate,
    };
  }

  componentDidUpdate(prevProps, prevState) {
    const { startDate, endDate } = this.state;
    const { startDate: prevStartDate, endDate: prevEndDate } = prevState;
    const { onChangeStartDate, onChangeEndDate } = this.props;

    if (!isEqual(startDate, prevStartDate) && onChangeStartDate) {
      onChangeStartDate(startDate);
    }

    if (!isEqual(endDate, prevEndDate) && onChangeEndDate) {
      onChangeEndDate(endDate);
    }
  }

  onMouseLeaveOfCalendar = () => this.setState({ hoveredDate: null });

  onMouseEnterOfDay = (hoveredDate) => {
    if (this.state.startDate && !this.state.endDate) {
      this.setState({ hoveredDate: hoveredDate.date });
    }
  }

  onDateSelected = ({ selectable, date }) => {
    if (!selectable) return;

    const { isSettingStartDate, isSettingEndDate } = this.state;

    if (isSettingStartDate) {
      this.selectStartDate(date);
    } else if (isSettingEndDate) {
      this.selectEndDate(date);
    } else if (!isSettingStartDate && !isSettingEndDate) {
      this.resetWithStartDate(date);
    }

    this.setState({
      hoveredDate: null,
    }, () => {
      const { startDate, endDate } = this.state;

      if (startDate && endDate) {
        this.props.onRangeSelected({ startDate, endDate });
      }
    });
  };

  selectStartDate = (startDate) => {
    const endDate = startDate <= this.state.endDate ? this.state.endDate : null;
    const endDateRequiresSelection = endDate === null;

    this.setState({
      startDate,
      endDate,
      isSettingStartDate: false,
      isSettingEndDate: endDateRequiresSelection,
    });
  };

  selectEndDate = (endDate) => {
    if (endDate <= this.state.startDate) return;

    this.setState({
      endDate,
      isSettingEndDate: false,
    });
  };

  resetWithStartDate = (startDate) => {
    this.setState({
      startDate,
      endDate: null,
      isSettingStartDate: false,
      isSettingEndDate: true,
    });
  };

  isInRange = (date) => {
    const {
      startDate, endDate, hoveredDate,
    } = this.state;

    if (!startDate || !endDate) return false;

    return isWithinRange(date, startDate, endDate || hoveredDate);
  }

  render() {
    const {
      monthNames,
      weekdayNames,
      monthsToDisplay,
      stacked,
      disabledDates,
      interactiveDisabledDates,
      ...rest
    } = this.props;

    const selectedDates = [this.state.startDate, this.state.endDate];

    return (
      <Dayzed
        {...rest}
        selected={selectedDates}
        monthsToDisplay={monthsToDisplay}
        onDateSelected={this.onDateSelected}
        render={({
            calendars,
            getBackProps,
            getForwardProps,
            getDateProps,
          }) => {
            if (!calendars.length) return null;

            return (
              <Box onMouseLeave={this.onMouseLeaveOfCalendar}>
                <CalendarNav
                  prevProps={getBackProps({ calendars })}
                  nextProps={getForwardProps({ calendars })}
                />

                <Flex flexWrap="wrap">
                  {calendars.map(calendar => (
                    <CalendarMonth
                      key={`${calendar.month}${calendar.year}`}
                      monthsToDisplay={monthsToDisplay}
                      monthName={monthNames[calendar.month]}
                      month={calendar.month}
                      year={calendar.year}
                      stacked={stacked}
                      weekdayNames={weekdayNames}
                      weeks={calendar.weeks}
                      getDateProps={getDateProps}
                      disabledDates={disabledDates}
                      interactiveDisabledDates={interactiveDisabledDates}
                      onMouseEnterOfDay={this.onMouseEnterOfDay}
                      isInRange={this.isInRange}
                    />
                  ))}
                </Flex>
              </Box>
            );
          }}
      />
    );
  }
}

DateRangePicker.defaultProps = {
  monthsToDisplay: 1,
  firstDayOfWeek: 1,
  stacked: false,
  minDate: subDays(new Date(), 1),
  disabledDates: [],
  monthNames: ['Jan', 'Feb', 'Mar', 'Apr', 'May', 'Jun', 'Jul', 'Aug', 'Sep', 'Oct', 'Nov', 'Dec'],
  weekdayNames: ['Mon', 'Tue', 'Wed', 'Thu', 'Fri', 'Sat', 'Sun'],
  interactiveDisabledDates: false,
  startDate: null,
  endDate: null,
  onChangeStartDate: null,
  onChangeEndDate: null,
  setStartDate: false,
  setEndDate: false,
};

DateRangePicker.propTypes = {
  monthsToDisplay: PropTypes.number,
  firstDayOfWeek: PropTypes.number,
  stacked: PropTypes.bool,
  minDate: PropTypes.instanceOf(Date),
  disabledDates: PropTypes.arrayOf(PropTypes.instanceOf(Date)),
  interactiveDisabledDates: PropTypes.bool,
  monthNames: PropTypes.arrayOf(PropTypes.string),
  weekdayNames: PropTypes.arrayOf(PropTypes.string),
  startDate: PropTypes.instanceOf(Date),
  endDate: PropTypes.instanceOf(Date),
  onChangeStartDate: PropTypes.func,
  onChangeEndDate: PropTypes.func,
  onRangeSelected: PropTypes.func.isRequired,
  setStartDate: PropTypes.bool,
  setEndDate: PropTypes.bool,
};

export default DateRangePicker;
