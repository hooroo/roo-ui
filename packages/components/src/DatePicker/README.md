# DatePicker

<!-- STORY -->

## Installation

```shell
$ yarn add @roo-ui/components
```

## Example

```js
import { DatePicker } from '@roo-ui/components';

export default (
  <DatePicker
    onDateSelected={console.log}
  />
);
```

## Properties

| Name                       | Description                          | Type      | Default | Required? |
|----------------------------|--------------------------------------|-----------|---------|-----------|
| `monthsToDisplay`          | Number of months to display          | `number`  | 1       | -         |
| `stacked`                  | Stack the calendar months vertically | `boolean` | false   | -         |
| `monthNames`               | Month names                          | `array`   | -       | -         |
| `weekdayNames`             | Weekday names                        | `array`   | -       | -         |
| `disabledDates`            | Dates which will be disabled         | `array`   | []      | -         |
| `interactiveDisabledDates` | Disabled dates will be clickable     | `boolean` | false   | -         |


## Customization

This component can be passed additional props defined in the [dayzed](https://github.com/deseretdigital/dayzed) documentation.
