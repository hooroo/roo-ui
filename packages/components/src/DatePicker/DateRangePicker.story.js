import React from 'react';
import { storiesOf } from '@storybook/react';
import { number, boolean } from '@storybook/addon-knobs';
import { withDocs } from 'storybook-readme';
import addDays from 'date-fns/add_days';
import subDays from 'date-fns/sub_days';

import DateRangePicker from './DateRangePicker';
import README from './DateRangePicker.README.md';

storiesOf('Components|DateRangePicker', module)
  .addDecorator(withDocs(README))
  .add('default', () => (
    <DateRangePicker
      monthsToDisplay={number('Months to display', 1)}
      minDate={subDays(new Date(), 10)}
      maxDate={addDays(new Date(), 60)}
      startDate={addDays(new Date(), 3)}
      endDate={addDays(new Date(), 7)}
      stacked={boolean('Stacked', false)}
      setStartDate={boolean('Change start date', false)}
      setEndDate={boolean('Change end date', false)}
      onRangeSelected={console.log} // eslint-disable-line no-console
      onChangeStartDate={console.log} // eslint-disable-line no-console
      onChangeEndDate={console.log} // eslint-disable-line no-console
    />
  ));
