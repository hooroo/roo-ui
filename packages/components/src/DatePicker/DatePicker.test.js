import React from 'react';
import { qantas as theme } from '@roo-ui/themes';
import { shallowWithTheme } from '@roo-ui/test-utils';

import DatePicker from '.';

describe('<DatePicker />', () => {
  let wrapper;
  let childrenWrapper;

  const props = {
    date: new Date('2018-07-03'),
    minDate: new Date('2018-06-01'),
    maxDate: new Date('2018-08-20'),
    selected: new Date('2018-07-20'),
    onDateSelected: jest.fn(),
    monthsToDisplay: 1,
    stacked: true,
    weekdayNames: ['S', 'M', 'T', 'W', 'T', 'F', 'S'],
    disabledDates: [new Date('2018-07-04'), new Date('2018-07-05')],
  };

  beforeEach(() => {
    wrapper = shallowWithTheme(<DatePicker {...props} />, theme);
    childrenWrapper = wrapper.dive();
  });

  it('renders correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('renders children correctly', () => {
    expect(childrenWrapper).toMatchSnapshot();
  });

  describe('<Dayzed />', () => {
    it('passes down props.monthsToDisplay', () => {
      expect(wrapper.find('Dayzed').prop('monthsToDisplay')).toEqual(props.monthsToDisplay);
    });

    it('assigns the rest of the props', () => {
      expect(wrapper.props()).toEqual(expect.objectContaining({
        date: props.date,
        minDate: props.minDate,
        maxDate: props.maxDate,
        selected: props.selected,
        onDateSelected: props.onDateSelected,
      }));
    });
  });

  describe('<CalendarMonth />', () => {
    it('passes the month name to each calendar month', () => {
      expect(childrenWrapper.find('CalendarMonth').prop('monthName')).toEqual('Jul');
    });

    it('passes down props.stacked', () => {
      expect(childrenWrapper.find('CalendarMonth').prop('stacked')).toEqual(props.stacked);
    });

    it('passes down props.monthsToDisplay', () => {
      expect(childrenWrapper.find('CalendarMonth').prop('monthsToDisplay')).toEqual(props.monthsToDisplay);
    });
  });
});
