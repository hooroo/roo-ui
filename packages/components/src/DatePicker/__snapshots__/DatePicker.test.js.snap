// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DatePicker /> renders children correctly 1`] = `
<Box
  position="relative"
>
  <CalendarNav
    nextProps={
      Object {
        "aria-label": "Go forward 1 month",
        "disabled": false,
        "onClick": [Function],
      }
    }
    prevProps={
      Object {
        "aria-label": "Go back 1 month",
        "disabled": false,
        "onClick": [Function],
      }
    }
  />
  <Box
    flexWrap="wrap"
  >
    <CalendarMonth
      disabledDates={
        Array [
          2018-07-04T00:00:00.000Z,
          2018-07-05T00:00:00.000Z,
        ]
      }
      getDateProps={[Function]}
      interactiveDisabledDates={false}
      key="62018"
      month={6}
      monthName="Jul"
      monthsToDisplay={1}
      stacked={true}
      weekdayNames={
        Array [
          "S",
          "M",
          "T",
          "W",
          "T",
          "F",
          "S",
        ]
      }
      weeks={
        Array [
          Array [
            "",
            "",
            "",
            "",
            "",
            "",
            Object {
              "date": 2018-06-30T14:00:00.000Z,
              "nextMonth": false,
              "prevMonth": false,
              "selectable": true,
              "selected": false,
              "today": false,
            },
          ],
          Array [
            Object {
              "date": 2018-07-01T14:00:00.000Z,
              "nextMonth": false,
              "prevMonth": false,
              "selectable": true,
              "selected": false,
              "today": false,
            },
            Object {
              "date": 2018-07-02T14:00:00.000Z,
              "nextMonth": false,
              "prevMonth": false,
              "selectable": true,
              "selected": false,
              "today": false,
            },
            Object {
              "date": 2018-07-03T14:00:00.000Z,
              "nextMonth": false,
              "prevMonth": false,
              "selectable": true,
              "selected": false,
              "today": false,
            },
            Object {
              "date": 2018-07-04T14:00:00.000Z,
              "nextMonth": false,
              "prevMonth": false,
              "selectable": true,
              "selected": false,
              "today": false,
            },
            Object {
              "date": 2018-07-05T14:00:00.000Z,
              "nextMonth": false,
              "prevMonth": false,
              "selectable": true,
              "selected": false,
              "today": false,
            },
            Object {
              "date": 2018-07-06T14:00:00.000Z,
              "nextMonth": false,
              "prevMonth": false,
              "selectable": true,
              "selected": false,
              "today": false,
            },
            Object {
              "date": 2018-07-07T14:00:00.000Z,
              "nextMonth": false,
              "prevMonth": false,
              "selectable": true,
              "selected": false,
              "today": false,
            },
          ],
          Array [
            Object {
              "date": 2018-07-08T14:00:00.000Z,
              "nextMonth": false,
              "prevMonth": false,
              "selectable": true,
              "selected": false,
              "today": false,
            },
            Object {
              "date": 2018-07-09T14:00:00.000Z,
              "nextMonth": false,
              "prevMonth": false,
              "selectable": true,
              "selected": false,
              "today": false,
            },
            Object {
              "date": 2018-07-10T14:00:00.000Z,
              "nextMonth": false,
              "prevMonth": false,
              "selectable": true,
              "selected": false,
              "today": false,
            },
            Object {
              "date": 2018-07-11T14:00:00.000Z,
              "nextMonth": false,
              "prevMonth": false,
              "selectable": true,
              "selected": false,
              "today": false,
            },
            Object {
              "date": 2018-07-12T14:00:00.000Z,
              "nextMonth": false,
              "prevMonth": false,
              "selectable": true,
              "selected": false,
              "today": false,
            },
            Object {
              "date": 2018-07-13T14:00:00.000Z,
              "nextMonth": false,
              "prevMonth": false,
              "selectable": true,
              "selected": false,
              "today": false,
            },
            Object {
              "date": 2018-07-14T14:00:00.000Z,
              "nextMonth": false,
              "prevMonth": false,
              "selectable": true,
              "selected": false,
              "today": false,
            },
          ],
          Array [
            Object {
              "date": 2018-07-15T14:00:00.000Z,
              "nextMonth": false,
              "prevMonth": false,
              "selectable": true,
              "selected": false,
              "today": false,
            },
            Object {
              "date": 2018-07-16T14:00:00.000Z,
              "nextMonth": false,
              "prevMonth": false,
              "selectable": true,
              "selected": false,
              "today": false,
            },
            Object {
              "date": 2018-07-17T14:00:00.000Z,
              "nextMonth": false,
              "prevMonth": false,
              "selectable": true,
              "selected": false,
              "today": false,
            },
            Object {
              "date": 2018-07-18T14:00:00.000Z,
              "nextMonth": false,
              "prevMonth": false,
              "selectable": true,
              "selected": false,
              "today": false,
            },
            Object {
              "date": 2018-07-19T14:00:00.000Z,
              "nextMonth": false,
              "prevMonth": false,
              "selectable": true,
              "selected": true,
              "today": false,
            },
            Object {
              "date": 2018-07-20T14:00:00.000Z,
              "nextMonth": false,
              "prevMonth": false,
              "selectable": true,
              "selected": false,
              "today": false,
            },
            Object {
              "date": 2018-07-21T14:00:00.000Z,
              "nextMonth": false,
              "prevMonth": false,
              "selectable": true,
              "selected": false,
              "today": false,
            },
          ],
          Array [
            Object {
              "date": 2018-07-22T14:00:00.000Z,
              "nextMonth": false,
              "prevMonth": false,
              "selectable": true,
              "selected": false,
              "today": false,
            },
            Object {
              "date": 2018-07-23T14:00:00.000Z,
              "nextMonth": false,
              "prevMonth": false,
              "selectable": true,
              "selected": false,
              "today": false,
            },
            Object {
              "date": 2018-07-24T14:00:00.000Z,
              "nextMonth": false,
              "prevMonth": false,
              "selectable": true,
              "selected": false,
              "today": false,
            },
            Object {
              "date": 2018-07-25T14:00:00.000Z,
              "nextMonth": false,
              "prevMonth": false,
              "selectable": true,
              "selected": false,
              "today": false,
            },
            Object {
              "date": 2018-07-26T14:00:00.000Z,
              "nextMonth": false,
              "prevMonth": false,
              "selectable": true,
              "selected": false,
              "today": false,
            },
            Object {
              "date": 2018-07-27T14:00:00.000Z,
              "nextMonth": false,
              "prevMonth": false,
              "selectable": true,
              "selected": false,
              "today": false,
            },
            Object {
              "date": 2018-07-28T14:00:00.000Z,
              "nextMonth": false,
              "prevMonth": false,
              "selectable": true,
              "selected": false,
              "today": false,
            },
          ],
          Array [
            Object {
              "date": 2018-07-29T14:00:00.000Z,
              "nextMonth": false,
              "prevMonth": false,
              "selectable": true,
              "selected": false,
              "today": false,
            },
            Object {
              "date": 2018-07-30T14:00:00.000Z,
              "nextMonth": false,
              "prevMonth": false,
              "selectable": true,
              "selected": false,
              "today": false,
            },
            "",
            "",
            "",
            "",
            "",
          ],
        ]
      }
      year={2018}
    />
  </Box>
</Box>
`;

exports[`<DatePicker /> renders correctly 1`] = `
<Dayzed
  date={2018-07-03T00:00:00.000Z}
  firstDayOfWeek={1}
  maxDate={2018-08-20T00:00:00.000Z}
  minDate={2018-06-01T00:00:00.000Z}
  monthsToDisplay={1}
  onDateSelected={[MockFunction]}
  onOffsetChanged={[Function]}
  render={[Function]}
  selected={2018-07-20T00:00:00.000Z}
  showOutsideDays={false}
/>
`;
